US-404006: Implement Query Builder Matcher

kusumaghoshdastidar opened the pull request:
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=404006

kusumaghoshdastidar committed:
[US-404006]: Implement Query Builder Matcher

kusumaghoshdastidar committed:
Implement Filter mapper

kusumaghoshdastidar committed:
Test Query Builder Matcher

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=CAFapi&REPO_NAME=querybuilder&BRANCH_NAME=US-404006) to go ahead and create the build...

kusumaghoshdastidar committed:
New line

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/CAFapi/job/CAFapi~querybuilder~US-404006~CI

kusumaghoshdastidar committed:
Tests for Long and String values

kusumaghoshdastidar committed:
Refactor code

kusumaghoshdastidar committed:
More string tests

kusumaghoshdastidar committed:
Fulltext tests

kusumaghoshdastidar committed:
Indent

kusumaghoshdastidar committed:
Refactor

kusumaghoshdastidar committed:
File path tests

kusumaghoshdastidar committed:
Fix file path tokenizing

kusumaghoshdastidar committed:
Cleanup TODOs

kusumaghoshdastidar committed:
Fix path separator

kusumaghoshdastidar committed:
Refactor number parsing

kusumaghoshdastidar committed:
Add print helper methods

kusumaghoshdastidar committed:
Resolve nullable warnings

kusumaghoshdastidar committed:
Resolve type param naming

kusumaghoshdastidar committed:
Format comments

kusumaghoshdastidar committed:
Use builtin function

kusumaghoshdastidar committed:
Update function visibility

kusumaghoshdastidar committed:
Drop line

kusumaghoshdastidar committed:
Remove TODO

kusumaghoshdastidar committed:
Leave comment in

kusumaghoshdastidar committed:
Refactor file path tokenization

kusumaghoshdastidar committed:
Refactor mapping function usage

kusumaghoshdastidar committed:
Fix nullable warnings

kusumaghoshdastidar committed:
Fix mapper namespace

kusumaghoshdastidar committed:
Update readme

kusumaghoshdastidar committed:
New line

kusumaghoshdastidar committed:
Usage document

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

kusumaghoshdastidar committed:
Update README.md

dermot-hardy committed:
Drop the BOM from the Tests project

dermot-hardy commented on README.md:
````diff
@@ -0,0 +1,330 @@
+        </sec-cyberres-nuget-prerelease>
+    </packageSourceCredentials>
+</configuration>
+```
````
We shouldn't reference internal company systems in open-source projects.  When this is released it will be published to [https://www.nuget.org/](https://www.nuget.org/) so no special NuGet configuration will be required.

dermot-hardy commented on README.md:
```diff
@@ -0,0 +1,330 @@
+    <PackageReference Include="MicroFocus.CafApi.QueryBuilder" Version="1.0.0-*" />
+    <PackageReference Include="MicroFocus.CafApi.QueryBuilder.Mapper" Version="1.0.0-*" />
+    <PackageReference Include="MicroFocus.CafApi.QueryBuilder.Matcher" Version="1.0.0-US-*" />
+    <PackageReference Include="MicroFocus.CafApi.QueryBuilder.Vqll.Parsers.SystemJson" Version="1.0.0-US-*" />
```
Just use `*` in the ReadMe to reference the latest version, whatever that may be.

```suggestion
    <PackageReference Include="MicroFocus.CafApi.QueryBuilder" Version="*" />
    <PackageReference Include="MicroFocus.CafApi.QueryBuilder.Mapper" Version="*" />
    <PackageReference Include="MicroFocus.CafApi.QueryBuilder.Matcher" Version="*" />
    <PackageReference Include="MicroFocus.CafApi.QueryBuilder.Vqll.Parsers.SystemJson" Version="*" />
```

dermot-hardy commented on README.md:
```diff
@@ -0,0 +1,330 @@
+to a `Filter<IMatcherFieldSpec>` using the `FilterMapper.Map` method.
+
+
+### Sample console application
```
These are too long and complete for a Readme.  Short fragments interspersed with explanatory text would be great but fully fleshed out runnable code should be in their own files, or an example project or something like that.

dermot-hardy commented on README.md:
```diff
@@ -0,0 +1,330 @@
+using MicroFocus.CafApi.QueryBuilder.Matcher;
+using MicroFocus.CafApi.QueryBuilder.Vqll.Parsers.SystemJson;
+
+namespace MicroFocus.Verity.QueryBuilderUsage
```
There should be no references to Verity in open-source.

dermot-hardy reviewed (changes requested):
Only reviewed ReadMe file so far:

kusumaghoshdastidar committed:
Update README.md

Co-authored-by: Dermot Hardy <dermot.hardy@microfocus.com>

kusumaghoshdastidar committed:
Refactor Mapper

kusumaghoshdastidar committed:
Simplify

kusumaghoshdastidar committed:
Simplify

kusumaghoshdastidar committed:
Merge branch 'main' into US-404006

kusumaghoshdastidar committed:
Simplify content

kusumaghoshdastidar committed:
Refactor mapping function

kusumaghoshdastidar committed:
Line wrap

dermot-hardy committed:
Made some minor whitespace changes

dermot-hardy committed:
Mapper: Use extension method for recursive calls

dermot-hardy committed:
General code review - minor changes

dermot-hardy commented on dotnet/src/MicroFocus.CafApi.QueryBuilder.Matcher/MatcherQueryVisitor.cs:
```diff
@@ -41,145 +44,358 @@ public bool GetResult()
+            if (fieldSpec.IsCaseInsensitive)
+            {
+                _isMatch = GetFieldValuesAsStream(fieldSpec)
+                    .Any(v => string.Equals(v, value, StringComparison.CurrentCultureIgnoreCase));
```
Using `StringComparison.CurrentCultureIgnoreCase` will cause different behavior on different platforms.  We should avoid anything that will make the code behave differently in different environments.

dermot-hardy commented on dotnet/src/MicroFocus.CafApi.QueryBuilder.Matcher/MatcherQueryVisitor.cs:
```diff
@@ -41,145 +44,358 @@ public bool GetResult()
         public void VisitContains(IMatcherFieldSpec<Document> fieldSpec, string value)
         {
-            throw new NotImplementedException();
+            string valueToCompare = fieldSpec.IsCaseInsensitive ? value.ToUpper(new CultureInfo("en-US")) : value;
```
We should use `.ToUpperInvariant()` for this.

dermot-hardy commented on dotnet/src/MicroFocus.CafApi.QueryBuilder.Matcher/MatcherQueryVisitor.cs:
```diff
@@ -41,145 +44,358 @@ public bool GetResult()
+            return GetStringValues(fieldSpec).Where(v =>
+            {
+                return long.TryParse(v, NumberStyles.AllowLeadingSign, null, out long xi);
+            }).Select(v => long.Parse(v));
```
With the `.TryParse()` you are passing `NumberStyles.AllowLeadingSign` which you are not on the `.Parse()`.  This could lead to a situation where the trial succeeds but the actual parse doesn't.  Really what you should do here is to use an [Anonymous Types](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/types/anonymous-types) to avoid the need for the second parse.

dermot-hardy committed:
Code review

- Removed platform specific code
- Used .ToUpperInvariant() rather than .ToUpper(new CultureInfo("en-US"))
- Avoided the double-parse
